### PR TITLE
Fix #4653: check method return type is value type

### LINKED ISSUE #4653: Crash after error
```scala
object Numbers {
  def test(numbers: Numbers): Unit = {
    import numbers._
    def succDeco(succ: Nat): SuccAPI = new SuccAPI[Int](???)
    succDeco(null.asInstanceOf[Nat]).pred
  }
  class SuccAPI[X](val nums: Numbers) {
     def pred: nums.Nat = ???
  }
}
trait Numbers {
  type Nat
}
```
crashes the compiler after an error is emitted

```scala
-- [E056] Syntax Error: tests/run/fully-abstract-nat-8.scala:6:29 --------------
6 |    def succDeco(succ: Nat): SuccAPI = new SuccAPI[Int](???)
  |                             ^^^^^^^
  |                             missing type parameter for Numbers.SuccAPI
assertion failure for Numbers.SuccAPI <:< ?{ pred: ? }, frozen = false
exception occurred while typechecking tests/run/fully-abstract-nat-8.scala
exception occurred while compiling tests/run/fully-abstract-nat-8.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: invalid prefix HKTypeLambda(List(X), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), AppliedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),module class Numbers$)),class SuccAPI),List(TypeParamRef(X))))
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.core.Types$NamedType.<init>(Types.scala:1592)
	at dotty.tools.dotc.core.Types$TermRef.<init>(Types.scala:2085)
	at dotty.tools.dotc.core.Types$CachedTermRef.<init>(Types.scala:2124)
	at dotty.tools.dotc.core.Uniques$NamedTypeUniques.newType$1(Uniques.scala:60)
	at dotty.tools.dotc.core.Uniques$NamedTypeUniques.enterIfNew(Uniques.scala:65)
	at dotty.tools.dotc.core.Types$TermRef$.apply(Types.scala:2155)
	at dotty.tools.dotc.core.Types$TermRef$.apply(Types.scala:2161)
	at dotty.tools.dotc.core.Types$NamedType$.apply(Types.scala:2147)
	at dotty.tools.dotc.core.Types$NamedType.reload$1(Types.scala:2035)
	at dotty.tools.dotc.core.Types$NamedType.withPrefix(Types.scala:2042)
	at dotty.tools.dotc.core.Types$NamedType.derivedSelect(Types.scala:2002)
	at dotty.tools.dotc.core.Types$TypeMap.derivedSelect(Types.scala:3925)
	at dotty.tools.dotc.core.Types$ApproximatingTypeMap.derivedSelect(Types.scala:4197)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$4(TypeOps.scala:61)
	at dotty.tools.dotc.reporting.trace$.op1$1(trace.scala:18)
	at dotty.tools.dotc.reporting.trace$.conditionally(trace.scala:20)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:54)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$5(TypeOps.scala:61)
	at dotty.tools.dotc.core.Types$VariantTraversal.atVariance(Types.scala:3911)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$4(TypeOps.scala:61)
	at dotty.tools.dotc.reporting.trace$.op1$1(trace.scala:18)
	at dotty.tools.dotc.reporting.trace$.conditionally(trace.scala:20)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:54)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:4013)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.$anonfun$apply$4(TypeOps.scala:67)
	at dotty.tools.dotc.reporting.trace$.op1$1(trace.scala:18)
	at dotty.tools.dotc.reporting.trace$.conditionally(trace.scala:20)
	at dotty.tools.dotc.core.TypeOps$AsSeenFromMap.apply(TypeOps.scala:54)
	at dotty.tools.dotc.core.TypeOps.asSeenFrom(TypeOps.scala:28)
	... 
```

### PR DESCRIPTION
**Pull request by Medowhill**

### CODE REVIEW

**Comment by liufengyun** [comment]
The following code:

```Scala
trait T {
  type S
}

class C {
  class D[X](val t: T) {
     def bar: t.S = ???
  }

  val foo: D = new D[Int](???) // error
  foo.bar
}
```

gets the following error:

```Scala
-- [E056] Syntax Error: examples/abc.scala:10:11 -------------------------------
10 |  val foo: D = new D[Int](???) // error
   |           ^
   |           missing type parameter for C.this.D
one error found
```

@Medowhill could you please try if we can produce similar error messages?

File: compiler/src/dotty/tools/dotc/typer/Namer.scala [hunk 1]
```diff
@@ -1153,7 +1153,12 @@ class Namer { typer: Typer =>
       case _ =>
         WildcardType
     }
-    paramFn(typedAheadType(mdef.tpt, tptProto).tpe)
+    val tp = typedAheadType(mdef.tpt, tptProto).tpe
+    paramFn(
+      if (tp.isValueType || (tp match {
+        case ExprType(tp1) => tp1.isValueType
+        case _ => false
+      })) tp else ErrorType("A type of a value or a return type of a method should be a value type."))
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/Namer.scala]: It's more succinct to write `tp.widenExpr.isValueType` here. I'm curious which test fails without the widening?

Comment (Medowhill) [reply compiler/src/dotty/tools/dotc/typer/Namer.scala]: I'll fix it. Thank you!
When a method has a by-name parameter, typing `ValDef` of the parameter results `ExprType`.

File: compiler/src/dotty/tools/dotc/typer/Namer.scala [hunk 2]
```diff
@@ -1153,7 +1153,9 @@ class Namer { typer: Typer =>
       case _ =>
         WildcardType
     }
-    paramFn(typedAheadType(mdef.tpt, tptProto).tpe)
+    val tp = typedAheadType(mdef.tpt, tptProto).tpe
+    paramFn(if (tp.widenExpr.isValueType) tp
+            else ErrorType("A type of a value or a return type of a method should be a value type."))
```
[the remark this answers is not in the document]
Comment (Medowhill) [reply compiler/src/dotty/tools/dotc/typer/Namer.scala]:
```diff
diff --git a/compiler/src/dotty/tools/dotc/typer/Namer.scala b/compiler/src/dotty/tools/dotc/typer/Namer.scala
index 8de871adbc3..2933c3dc92a 100644
--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -1153,9 +1153,7 @@ class Namer { typer: Typer =>
       case _ =>
         WildcardType
     }
-    val tp = typedAheadType(mdef.tpt, tptProto).tpe
-    paramFn(if (tp.widenExpr.isValueType) tp
-            else ErrorType("A type of a value or a return type of a method should be a value type."))
+    paramFn(checkSimpleKinded(typedAheadType(mdef.tpt, tptProto)).tpe)
   }
```
It looks clearer. Thank you!

**Pull request merged**